Allow caching of the logout route

rullzer opened the pull request:
It is a bit more general. If the only argument is the requesttoken. So a
get request will work with the CSRF protection. The route will be
generated without the parameters and the request token will be added at
the end.

Before the route could not properly be cached as the request token was
always changing. However now it can be cached saving precious cpu
cycles.

Noticed while looking at https://github.com/nextcloud/server/pull/13712. While https://github.com/nextcloud/server/pull/13712 is a general improvement for route generation this fixes the non caching of this specific route.

MorrisJobke commented:
I really don't like that this is handled in there. I would handle it directly in the logout URL handling - just request the URL without any parameter and manually add it then. This reduces unnecessary complexity in the reused method `generate()` and makes it easier to understand, because no magic like this is handled.

rullzer commented:
Then somebody with JS magic skills needs to tackle it

MorrisJobke commented:
> Then somebody with JS magic skills needs to tackle it

Why? It's PHP as well - I would just do it in here:

https://github.com/nextcloud/server/blob/bb86a8ca36a1aeb73a54339340b0d1ddc2f249c3/lib/private/legacy/user.php#L266-L280

😉

rullzer commented:
Aaaaah right ok makes sense. Let me do that then...

rullzer committed:
Request plain logout url

By requesting the plain logout url we allow it to be properly cached by
the caching router. We just add the requesttoken manually.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

nickvergessen reviewed (approved):
:see_no_evil:

ChristophWurst reviewed (approved):
:+1: :elephant: :see_no_evil:

MorrisJobke commented:
Still failing:

```
--- Expected
+++ Actual
@@ @@
-        'href' => 'https://example.com/logout'
+        'href' => 'https://example.com/logout?requesttoken=Wq1opm+k0/NnQAydww7painntbJdUuM6gyTGbzqAOUk=:NccK4zbBgdxWcViksE2QK1+OgdksEYBW1XCVNXTZUR4='
```

danxuliu commented:
The acceptance test failures seem legit; for me, logging out now (sometimes) results in an error page with
```
Access forbidden
CSRF check failed
```

MorrisJobke commented:
😢

rullzer commented:
> The acceptance test failures seem legit; for me, logging out now (sometimes) results in an error page with
> 
> ```
> Access forbidden
> CSRF check failed
> ```

Aaaah sometimes yes... OK I know what is going on... fix incomming
